### ISSUE TITLE
Call Stack Exceeded Error in isEqual

### ISSUE THREAD

**Issue by calebeby** [apollographql/apollo-client#4824]
👋 Thanks for apollo, it's pretty cool

**Intended outcome:**

Data fetches without error

**Actual outcome:**

Call Stack Exceeded Error

<img width="1235" alt="Screen Shot 2019-05-16 at 10 59 01 AM" src="https://user-images.githubusercontent.com/13206945/57876208-a9a7f780-77c9-11e9-8b92-8419dcaf9206.png">

I did some digging into this error, and here is what I found:

`isDifferentFromLastResult` is called here:
https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/QueryManager.ts#L697

`isEqual` is called here:
https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/ObservableQuery.ts#L275

The two objects that are passed to isEqual look like this:

<img width="898" alt="Screen Shot 2019-05-16 at 11 07 36 AM" src="https://user-images.githubusercontent.com/13206945/57876711-df99ab80-77ca-11e9-98ed-65929c364cc6.png">

`a` looks like:

![image](https://user-images.githubusercontent.com/13206945/57876791-1243a400-77cb-11e9-9ac9-4afa0fccfed4.png)

`b` looks like:

![image](https://user-images.githubusercontent.com/13206945/57876889-52a32200-77cb-11e9-8caa-3de19012a91f.png)

When you invoke the getters on `b` it looks like `a`. Both `b` and `a` are infinitely nested:

`a.allLinks[0].category.links[0]` is the same object as `a.allLinks[0]`
`b.allLinks[0].category.links[0]` is the same object as `b.allLinks[0]`

The way `isEqual` is written, it does not support infinitely nested objects. This leads to the maximum call stack error because `isEqual` keeps calling itself with properties deeper and deeper into the infinitely nested tree.

Why is `isEqual` passed an infinitely nested tree? Is `isEqual` supposed to support infinitely nested trees?

**How to reproduce the issue:**

I'm not going to create a minimal reproduction quite yet, first I wanted to check and see if you had any insight based on the debugging above ☝️

**Versions**

```
  System:
    OS: macOS 10.14.2
  Binaries:
    Node: 10.11.0 - /usr/local/bin/node
    Yarn: 1.15.2 - /usr/local/bin/yarn
    npm: 6.9.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 74.0.3729.131
    Firefox: 65.0
    Safari: 12.0.2
```

We are using `@nuxtjs/apollo@4.0.0-rc4` and `apollo-client@2.5.1`, although I don't think the error comes from the nuxt integration. We also are using `apollo-cache-inmemory@1.5.1`

Any help or further guidance in debugging would be appreciated 😄

**Comment by hammadj** [comment]
Verified this issue is still present in `apollo-client@2.6.0`. Replacing the apollo implementation of `isEqual` with lodash isEqual fixes the error. Infinitely nested trees should definitely be supported.

**Comment by benjamn** [comment]
This should (soon) be fixed by https://github.com/apollographql/apollo-client/pull/4915.

**Comment by benjamn** [comment]
This should now be fixed if you update to `apollo-utilities@1.3.1`, thanks to @capaj's PR #4903.

**Comment by Paul-Hebert** [comment]
Awesome, thanks @benjamn , @capaj !
